src/tools/data_collection_util.cpp

```diff
@@ -0,0 +1,620 @@
+// Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+// Axom includes
+#include "axom/config.hpp"
+#include "axom/core.hpp"
+#include "axom/slic.hpp"
+#include "axom/sidre.hpp"
+#include "axom/primal.hpp"
+
+#include "fmt/fmt.hpp"
+#include "CLI11/CLI11.hpp"
+
+// MFEM is required
+#ifndef AXOM_USE_MFEM
+  #error This example requires an Axom configured with MFEM
+#endif
+#include "mfem.hpp"
+
+// MPI is optional
+#ifdef AXOM_USE_MPI
+  #include "mpi.h"
+#endif
+
+namespace slic = axom::slic;
+namespace sidre = axom::sidre;
+namespace primal = axom::primal;
+
+//------------------------------------------------------------------------------
+
+/// Struct to help choose if we're loading a box (Cartesian) mesh or a file
+enum class MeshForm : int
+{
+  Box,
+  File
+};
+
+/// Struct to parse and store the input parameters
+struct Input
+{
+public:
+  std::string dcName;
+
+  int uniformRefinements {0};
+  int polynomialOrder {2};
+
+  MeshForm meshForm {MeshForm::Box};
+
+  // Box mesh parameters
+  std::vector<double> boxMins;
+  std::vector<double> boxMaxs;
+  std::vector<int> boxResolution;
+  int boxDim {-1};
+
+  // File mesh parameters
+  std::string mfemFile;
+  std::vector<double> fileScale;
+  std::vector<double> fileTranslate;
+
+private:
+  bool m_verboseOutput {false};
+
+  /**
+   * Fixes up parameters associated with a box mesh based on user-provided input
+   */
+  void fixBoxParams()
+  {
+    const bool dimProvided = boxDim > 0;
+    const bool rangeProvided = !(boxMins.empty() && boxMaxs.empty());
+    const bool resProvided = !boxResolution.empty();
+
+    // First check: we need at least one of: dimension, mins/maxs, resolution
+    if(!(dimProvided || rangeProvided || resProvided))
+    {
+      throw "Box mesh must have at least one of: dimension, mins and maxs or "
+        "resolution to determine the dimension";
+    }
+
+    int szMins = boxMins.size();
+    int szMaxs = boxMaxs.size();
+    int szRes = boxResolution.size();
+
+    // Error checking on provided dimension
+    if(dimProvided)
+    {
+      if(boxDim < 1 || boxDim > 3)
+      {
+        throw "Dimension cannot exceed 3";
```

Comment:
Error is confusing if you end up with boxDim == 0

Reply:
Agreed.

TODO:
- [ ] Fix this